README.md
# Galaxy: an asynchronous distributed parallel framework

`master`: [![Build Status](https://travis-ci.org/TACC/Galaxy.svg?branch=master)](https://travis-ci.org/TACC/Galaxy)
`dev`: [![Build Status](https://travis-ci.org/TACC/Galaxy.svg?branch=dev)](https://travis-ci.org/TACC/Galaxy)

Galaxy is an asynchronous parallel visualization framework for ray-based operations, including ray tracing, sampling and volume integration. Galaxy has been designed for performance in distributed computing environments, including in situ use. Galaxy builds upon [Intel OSPRay][1] and [Intel Embree][2] and includes ray queueing and sending logic inspired by [TACC GraviT][3]. 

Galaxy has been developed primarily by Greg Abram and Paul Navratil with funding from the US Department of Energy's Office of Science ASCR program (Drs. Lucy Nowell and Laura Biven, program managers) through sub-award from Los Alamos National Laboratory; from Intel Corporation through Intel Parallel Computing Center awards; and from the Texas Advanced Computing Center at the University of Texas at Austin.
A full list of contributors can be found in the `CONTRIBUTORS` file.

Galaxy is licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. A copy of the License is included with this software in the file `LICENSE`. If your copy does not contain the License, you may obtain a copy of the License at: [Apache License Version 2.0][15]. 
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.  

Copyright (c) 2014-2020 The University of Texas at Austin. All rights reserved.

## Galaxy dependencies
Galaxy has two types of dependencies: (1) components assumed to be already installed in an accessible spot; and (2) components that are associated with the Galaxy repository in the `third-party` subdirectory.

Galaxy assumes the following are already installed on your system in an accessible spot (these should all be available via package manager for your OS):
  * a modern MPI (Message Passing Interface) implementation
  * [a recent VTK][4] (at least version 7.x)
  * [Intel Thread Building Blocks (TBB)][6]
  * [PNG image format library][7]
  * [zlib compression library][8]
  * [CMake cross-platform build utility][9]
  * [GLFW][14] (required by OSPRay build)
  * [Python][18] (required for testing and using python scripts in `scripts`)
  * [Qt5][20] (at least v5.9, tested against v5.14)

Galaxy has the following components associated in the `third-party` subdirectory:
  * [Intel ISPC][10]
  * [Intel Embree][12]
  * [Intel OSPRay][13]
  * [rapidjson][16]

*NOTE:* the third-party component locations are intended for development builds of Galaxy. If you intend to install Galaxy for general use, you should install Embree, OSPRay and rapidjson into a generally-accessible location and use those locations in the Galaxy build.

## Installing Galaxy associated dependencies
Prior to building Galaxy itself, you should ensure that all the general dependencies are installed (we recommend your humble OS package manager). Once those are in place, you're ready to install the dependencies in `third-party`. There are two steps to this process: 1) downloading and patching the submodules, and 2) building and installing the submodules.

### Scripted build
We recommend using the Galaxy-provided build script `scripts/install-third-party.sh`, which will configure and build all submodules with a single command (if this script does not work for your configuration, please open a [GitHub issue][19]). If you need to customize your build, please see the next section.

To **download and build** the submodules: `cd` to the root directory of your local Galaxy repository and run the script `./scripts/install-third-party.sh`, which will (1) init and update the git submodules, (2) download ispc to where Galaxy expects to find it, and (3) build each in a `build` subdirectory and install into an `install` subdirectory. This script will populate the `third-party/ispc`, `third-party/embree`, `third-party/ospray`, and `third-pary/rapidjson` subdirectories. Once built, each third-party component will be installed to a single centralized location in `third-party/install`.

### Custom build
The third-party dependency build can also be done by hand, for example, to 
specify custom software locations or to change the configuration of Embree or OSPRay. Note that even with a custom build, the submodules still should be patched as described below.

To **download and patch** the submodules: `cd` to the root directory of your local Galaxy repository and run the script `./scripts/prep-third-party.sh`, which will (1) init and update the git submodules, and (2) download ispc to where Galaxy expects to find it. This script will populate the `third-party/ispc`, `third-party/embree`, `third-party/ospray`, and `third-pary/rapidjson` subdirectories.

To **build and install** the submodules, do the following in `third-party/embree`, then `third-part/ospray`, then `third-party/rapidjson` (`third-party/ispc` is already installed via the script):

```bash
mkdir build
cd build
cmake <cmake flags> .. && make && make install
```

*NOTE*: for the CMake flags used for each third-party library, see the CMAKE_FLAGS variables defined in `scripts/install-third-party.sh`.

#### Local installs of Galaxy associated dependencies
If you prefer to use local installs of any of the dependencies, you can follow instructions below. Make sure to issue `git submodule init <path>` and `git submodule update <path>` for the third-party dependencies you do *NOT* have locally installed. For example, to use the rapidjson submodule, type:
```bash
git submodule init third-party/rapidjson
git submodule update third-party/rapidjson
```

#### ISPC
The `third-party/ispc` directory contains a script to download a recent ISPC binary for MacOS or Linux. ISPC installed using this script should be detected automatically by the Galaxy CMake configuration. If you already have a recent ISPC installed (at least version 1.9.1) you are free to use it, though you might need to specify its location by hand in the CMake configurations for Embree, OSPRay, and Galaxy.

From the root directory of your Galaxy repository, issue the following commands:
```bash
cd third-party/ispc
./get-ispc.sh 
```

This will download ISPC (currently version 1.12.0) into `third-party/ispc/ispc-v<version>-<OS type>` (e.g., ispc-v1.10.0-macOS), and it will install the `ispc` executable in `third-party/install/bin`. If this binary does not work on your system, you will need to build ISPC by hand following the directions at the [ISPC website][10]. 

#### Embree
After updating the git submodules as described above, the `third-party/embree` directory should contain the Embree source tree. We recommend building in `third-party/embree/build` and installing to `third-party/install`, as doing so should enable Galaxy to find Embree. The recommended install directory is configured as part of the Embree cmake flags in `scripts/install-third-party.sh`.

First, create a build directory and build Embree. Assuming CMake can find all required dependencies, you can use the `cmake` command to configure the makefile for the Embree build:
```bash
mkdir build
cd build
cmake <cmake flags> .. && make && make install
```
If cmake complains about missing dependencies, you can specify or change their locations using cmake `-D<CMAKE VAR>` command-line syntax or using the interactive `ccmake` interface with `ccmake ..` in the build directory. See `scripts/install-third-party.sh` for recommended cmake flags.

#### OSPRay
Before installing OSPRay, make sure you have updated the Galaxy git submodules and successfully built Embree, as described above. Once the git submodules have been updated, the `third-party/ospray` directory should contain the OSPRay source tree. We recommend building in `third-party/ospray/build` and installing to `third-party/install`, as doing so should enable Galaxy to find OSPRay automatically. The recommended install directory is configured as part of the OSPRay cmake flags in `scripts/install-third-party.sh`.

First, create a build directory and build OSPRay. Assuming CMake can find all required dependencies, you can use the `cmake` command to configure the makefile for the OSPRay build:
```bash
mkdir build
cd build
cmake <cmake flags> .. && make && make install
```
If cmake complains about missing dependencies, you can specify or change their locations using cmake `-D<CMAKE VAR>` command-line syntax or using the interactive `ccmake` interface with `ccmake ..` in the build directory. See `scripts/install-third-party.sh` for recommended cmake flags.

#### rapidjson
Before installing rapidjson, make sure you have updated the Galaxy git submodules as described above. Once the git submodules have been updated, the `third-party/rapidjson` directory should contain the rapidjson source tree. We recommend building in `third-party/rapidjson/build` and installing to `third-party/install`, as doing so should enable Galaxy to find rapidjson automatically. The recommended install directory is configured as part of the RapidJSON cmake flags in `scripts/install-third-party.sh`.

First, create a build directory and build rapidjson. Assuming CMake can find all required dependencies, you can use the `cmake` command to configure the makefile for the rapidjson build:

```bash
mkdir build
cd build
cmake <cmake flags> .. && make && make install
```
If cmake complains about missing dependencies, you can specify or change their locations using cmake `-D<CMAKE VAR>` command-line syntax or using the interactive `ccmake` interface with `ccmake ..` in the build directory. See `scripts/install-third-party.sh` for recommended cmake flags.

## Building Galaxy
Before building Galaxy, make sure all assumed and third-party subdirectory dependencies have been installed as described above, which will allow the Galaxy CMake configuration to find all dependencies automatically. We recommend building in `<Galaxy root>/build` and installing to `<Galaxy root>/install`. The recommended install directory is configured as part of the Galaxy CMake configuration.

Assuming CMake can find all required dependencies, you can use the `cmake` command to configure the makefile for the Galaxy build:

```bash
mkdir build
cd build
cmake .. && make && make install
```
If cmake complains about missing dependencies, you can specify or change their locations using cmake `-D<CMAKE VAR>` command-line syntax or using the interactive `ccmake` interface with `ccmake ..`.

## Running Galaxy

There are several ways to run Galaxy.   In this document we currently only describe the batch mode applications; interactive viewers will be added to the documentation as time allows.

### Setting up the environment

In the install root, `galaxy.env` contains the necessary environment variable settings for running Galaxy.   If you use the bash or sh shells, you can 'source' that file; for csh you'll need to modify it.

We currently assume that your environment has been set up to find both VTK and Intel OSPRay libraries.   

### Preparing test data

```bash
mkdir $PROJECT/test
cd $PROJECT/test
```

Make some volumetric test data

`radial -r 256 256 256`

will create a .VTI dataset (radial-0.vti - note that radial will create timestep data if requested and the '-0' reflects the timestep) with several variables: ramps in X, Y and Z; oneBall is f(x, y, z) = mag(x, y, z) and eightBalls is f(x, y, z) = mag(abs(x) - 0.5, abs(y) - 0.5, abs(z) - 0.5)    We can create native format datasets:

`vti2vol radial-0.vti oneBall eightBalls`

will create radial-0-oneBall.vol, radial-0-eightBalls.vol and corresponding …raw files that actually contain the data as bricks of floats.

### Sample Galaxy State File

Galaxy uses a JSON state file format to describe data and visualization operations. We discuss a sample Galaxy configuraiton file below. For more details about Galaxy state files, see `docs/state_files.md`.

The following is a state file that should work with the datasets we created above, and will result in six renderings: two visualizations rendered from each of the three cameras.

```json
{
    "Datasets":
    [
      {
        "name": "oneBall",
        "type": "Volume",
        "filename": "radial-0-oneBall.vol"
      },
      {
        "name": "eightBalls",
        "type": "Volume",
        "filename": "radial-0-eightBalls.vol"
      }
    ],
    "Renderer": {
        "Lighting": {
          "Sources": [[1, 1, 0, 1]],
          "shadows": true,
          "Ka": 0.4,
          "Kd": 0.6,
          "ao count": 16,
          "ao radius": 0.5
      }
    },
    "Visualizations":
    [
      {
        "annotation": "",
        "operators":
          [
            {
              "type": "Volume",
              "dataset": "oneBall",
              "isovalues": [ 0.25 ],
              "volume rendering": false,
              "colormap": [
                [0.00,1.0,0.5,0.5],
                [0.25,0.5,1.0,0.5],
                [0.50,0.5,0.5,1.0],
                [0.75,1.0,1.0,0.5],
                [1.00,1.0,0.5,1.0]
               ],
              "slices": [ [1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0] ],
              "opacitymap": [
                [ 0.00, 0.05],
                [ 0.20, 0.05],
                [ 0.21, 0.00],
                [ 1.00, 0.00]
               ]
            }
          ]
      },
      {
        "annotation": "",
        "operators":
          [
            {
              "type": "Volume",
              "dataset": "eightBalls",
              "slices": [ [1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0] ],
              "volume rendering": true,
              "colormap": [
                [0.00,1.0,0.5,0.5],
                [0.25,0.5,1.0,0.5],
                [0.50,0.5,0.5,1.0],
                [0.75,1.0,1.0,0.5],
                [1.00,1.0,0.5,1.0]
               ],
              "opacitymap": [
                [ 0.00, 0.05],
                [ 0.20, 0.05],
                [ 0.21, 0.00],
                [ 1.00, 0.00]
               ]
            }
          ]
       }
    ],
    "Cameras":
    [
      {
        "aov": 30,
        "Xaov": 20,
        "viewpoint": [1, 2, -3],
        "viewcenter": [0, 0, 0],
        "viewup": [0, 1, 0]
      },
      {
        "aov": 30,
        "Xaov": 20,
        "viewpoint": [2, 1, -3],
        "viewcenter": [0, 0, 0],
        "viewup": [0, 1, 0]
      },
      {
        "aov": 30,
        "Xaov": 20,
        "viewpoint": [3, 1, -2],
        "viewcenter": [0, 0, 0],
        "viewup": [0, 1, 0]
      }

    ]
}
```

Considering each section in turn:

The Datasets section define the datasets that are available for visualization.   Each dataset specification includes the means to access the data, either using a file name or the information necessary to attach to an external source of data, and a name to be used internally.

The Renderer section includes the properties of the rendering, which are currently common among all the results of the run.   Rendering properties currently are simply the lighting model to be used, including the light sources themselves, whether to cast shadow rays or to add a fixed proportion of diffuse lighting, and whether to cast AO rays or to add a fixed proportion of ambient light.

The Visualizations section is an array, where each element (a visualization) contains an array of operators: one or more datasets and properties to be included in the visualization.   As an example, if the following is an element in a visualization operator array, that array will include the eightBalls dataset, with one slice, one isovalue and with volume rendering using the given transfer function.

```json
{
    "type": "Volume",
    "dataset": "eightBalls",
    "volume rendering": true,
    "slices": [ [0, 0, 1, 0] ],

RELEASE_NOTES.md
# Galaxy Release Notes

## version 0.5.0 - pending

* *TODO* - migrate volume rendering from OSPRay to OpenVKL[5]
* *TODO* - add OpenImageDenoise cleaning for asynchronous rendering[6]
* *TODO* - centralize data filtering codes currently in `src` to `src/filters`
* *TODO* - scripts to launch singularity-based multi-node distributed testing on TACC systems

## version 0.4.0 - 20200630
* Added `gui` and `guiserver`
  - a Qt-based node editor[4] graphical user interface for Galaxy workflows
  - added `nodeeditor` submodule
  - added Qt5 dependency
* Added `insitu` socket connectivity for in situ workflows
* Added `tracer` streamline generation and rendering
* Added `sampler` methods
  - Gradient-based sampling
  - Isovalue-based sampling
* Added demonstration codes for new features in `apps` 
* Updated to python3
  - python scripts
  - travis-ci config
  - travis-ci VTK-8.1 build

## version 0.3.2 - 20191212

* updated OSPRay version to 1.8.5
* updated Embree version to 3.6.1
* updated ISPC version to 1.12.0
* centralized third-party installation to third-party/install directory
* removed patchfiles for third-party packages, moved config options to cmake arguments in `scripts/install-third-party.sh`

## version 0.3.1 - 20190930

* added examples for Schlieren renderer in `examples/schlieren`
* added examples for various data samplers in `examples/sampletrace`
* added second, energy-accumulation schlieren algorithm
* added 'data.state' argument to `sampletrace` to track data
* moved image resolution specification to Camera
* added Cinema database example of Water in the Universe data in `examples/WitU`
* added 'data range' to visualizations to select range of interest

## version 0.3.0 - 20190821

* Added `sampler` with a variety of data-space sampling algorithms:
  - metropolis-hastings sampling as a multiserver tool 
  - ray-based sampling using the 'Sampler' infrastructure
    + Sampler is a subclass of Renderer that supports integration with active-Vis operators
    + active-Vis operators to implement sampling based on isosurface crossing and significant changes to gradient
  - interpolation of a scalar field onto a set of samples (as a multiserver tool)
* Support for vector-valued volumes
  - added JSON format for volume description files that implements 'number of components' field
* Initial implementation of RungeKutta operator to trace streamlines in vector volumes
  - also 'TraceToPathLines' operator to convert RungeKutta results to path lines for rendering
* Added `schlieren` to provide multi-hued Schlieren and shadowgraph-style rendering
  - also implemented 'float-image' output (in FITS format)
* Added `ospray` modules for class overrides to the underlying [OSPRay][1] rendering engine
* Separated data-generation and data-management tools into `data` from `apps`
* Added Docker containers:
    - [`galaxy`][2]: contains pre-built version of galaxy in CentOS linux
    - [`galaxy-base`][3]: contains pre-requisites for galaxy build in CentOS linux
* Added unit testing framework
* Renamed `async` app and associated source to `gxyviewer`
* Renamed `vis` app and associated source to `gxywriter`
* Renamed `mh` app and associated source to `mhwriter` 
* Renamed `amh` app and associated source to `mhviewer`
* updated third-party/ispc to download v1.10.0
* updated third-party/embree to use v3.5.2
* updated third-party/ospray to use v1.7.3

## version 0.2.1 - 20190325

* Moved Doxygen to root Galaxy directory
* Fixed includes for Fedora 29 build

## version 0.2.0 - 20190208

Refactorization of Galaxy core to cleanly partition framework, data, and renderer layers. Initial release of multiserver modularization of Galaxy framework.

* Added new `multiserver` component that enables Galaxy to live as a service (e.g. alongside a simulation *in situ*) and process requests that arrive via remote connections.
* Removed OSPRay and Embree dependencies from framework and data layers.
* Install now generates a GalaxyConfig.cmake file so that Galaxy components can be included in other programs
* Added Travis-CI build and test integration
* Added sanity-check gold image tests in `tests`
* Added Galaxy state file documentation in `doc/configuration_files.md`
* Added Python as a prerequisite for testing and use of python scripts in `scripts`

## version 0.1.0 - 20181005

Initial Galaxy release!

* Asynchronous tasking and communication framework
* Data layer for visualization and rendering
* Distributed ray tracing renderer built on Intel OSPRay and Embree for volume integration and geometry intersection, respectively
* Doxygen-based documentation of classes
* CMake build and packaging


[1]: https://ospray.org/
[2]: https://hub.docker.com/r/pnav/galaxy
[3]: https://hub.docker.com/r/pnav/galaxy-base
[4]: https://github.com/paceholder/nodeeditor
[5]: https://www.openvkl.org/
[6]: https://www.openimagedenoise.org/




src/apps/mh.cpp
// ========================================================================== //
// Copyright (c) 2014-2019 The University of Texas at Austin.                 //
// All rights reserved.                                                       //
//                                                                            //
// Licensed under the Apache License, Version 2.0 (the "License");            //
// you may not use this file except in compliance with the License.           //
// A copy of the License is included with this software in the file LICENSE.  //
// If your copy does not contain the License, you may obtain a copy of the    //
// License at:                                                                //
//                                                                            //
//     https://www.apache.org/licenses/LICENSE-2.0                            //
//                                                                            //
// Unless required by applicable law or agreed to in writing, software        //
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT  //
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.           //
// See the License for the specific language governing permissions and        //
// limitations under the License.                                             //
//                                                                            //
// ========================================================================== //

#include <iostream>

#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <cstdlib>
#include <math.h>
#include <dlfcn.h>

#include <dtypes.h>
#include <Application.h>
#include "Renderer.h"
#include "RenderingSet.h"

#include <ospray/ospray.h>

#include <time.h>
#include <boost/random/normal_distribution.hpp>
#include <boost/random/mersenne_twister.hpp>
#include <boost/random/variate_generator.hpp>

using namespace boost;

int mpiRank, mpiSize;

#include "Debug.h"

using namespace gxy;
using namespace std;

int total_samples = 100;

#define WIDTH   1920
#define HEIGHT 1080

int width  = WIDTH;
int height = HEIGHT;

float radius = 0.02;
float target = 0.0;
float power = 10.0;
float sigma = 1.0;
int   max_consecutive_misses = 10;
int   iteration_limit = 20000;
int   hot_start = 1000;
float scaling = 1.0;
int   skip = 10;

#define TF_NONE     0
#define TF_LINEAR   1
#define TF_GAUSSIAN 2


int tf_type = TF_NONE;
float tf0, tf1;

variate_generator<mt19937, normal_distribution<> > *generator;

void normally_distributed_vector(double* v)
{
  for (size_t i = 0; i < 3; ++i)
    v[i]=(*generator)();
}

float gaussian(float x, float m, float s)
{
  return ( 1 / ( s * sqrt(2*M_PI) ) ) * exp( -0.5 * pow( (x-m)/s, 2.0 ) );
}

#define RNDM ((float)rand() / RAND_MAX) 

class SampleMsg : public Work
{
    struct args
    {
        Key   vk;
        Key   pk;
        int   tf_type;
        float tf0;
        float tf1;
        float step;
        int   n_iterations;
        int   n_start;
        int   n_skip;
        int   n_miss;
    };

public:
  SampleMsg(VolumeP v, ParticlesP p, int tf_type, float tf0, float tf1, float step, int niterations, int nstart, int nskip, int nmiss)
            : SampleMsg(sizeof(args))
  {
        args *a = (args *)contents->get();
        a->vk = v->getkey();
        a->pk = p->getkey();
        a->tf_type = tf_type;
        a->tf0 = tf0;
        a->tf1 = tf1;
        a->step = step;
        a->n_iterations = niterations;
        a->n_start = nstart;
        a->n_skip = nskip;
        a->n_miss = nmiss;
    }

  ~SampleMsg() {}

  WORK_CLASS(SampleMsg, true)

private: 
  float istep, jstep, kstep;

    vec3f get_starting_point(VolumeP v)
    {
        Box *box = v->get_local_box();
        return vec3f(box->xyz_min.x + RNDM*(box->xyz_max.x - box->xyz_min.x),
                     box->xyz_min.y + RNDM*(box->xyz_max.y - box->xyz_min.y),
                     box->xyz_min.z + RNDM*(box->xyz_max.z - box->xyz_min.z));
    }

  float Q(VolumeP v, float s, args *a)
  {
    float q;

    if (a->tf_type == TF_LINEAR)
    {
      if (a->tf0 < a->tf1)
      {
        if (s < a->tf0) q = 0.0;
        else if (s > a->tf1) q = 1.0;
        else q = (s - a->tf0) / (a->tf1 - a->tf0);
      }
      else
      {
        if (s < a->tf1) q = 1.0;
        else if (s > a->tf0) q = 0.0;
        else q = (s - a->tf0) / (a->tf1 - a->tf0);
      }
    }
    else
    {
      q = gaussian(s, a->tf0, a->tf1);
    }

    return q;
  }

  float sample(VolumeP v, Particle& p) { return sample(v, p.xyz.x, p.xyz.y, p.xyz.z); }
  float sample(VolumeP v, vec3f xyz) { return sample(v, xyz.x, xyz.y, xyz.z); }
  float sample(VolumeP v, float x, float y, float z)
  {
    float dx, dy, dz;
    v->get_deltas(dx, dy, dz);

    float ox, oy, oz;
    v->get_ghosted_local_origin(ox, oy, oz);

    x = (x - ox) / dx;
    y = (y - oy) / dy;
    z = (z - oz) / dz;

    int ix = (int)x;
    int iy = (int)y;
    int iz = (int)z;

    dx = x - ix;
    dy = y - iy;
    dz = z - iz;

    int v000 = (ix + 0) * istep + (iy + 0) * jstep + (iz + 0) * kstep;
    int v001 = (ix + 0) * istep + (iy + 0) * jstep + (iz + 1) * kstep;
    int v010 = (ix + 0) * istep + (iy + 1) * jstep + (iz + 0) * kstep;
    int v011 = (ix + 0) * istep + (iy + 1) * jstep + (iz + 1) * kstep;
    int v100 = (ix + 1) * istep + (iy + 0) * jstep + (iz + 0) * kstep;
    int v101 = (ix + 1) * istep + (iy + 0) * jstep + (iz + 1) * kstep;
    int v110 = (ix + 1) * istep + (iy + 1) * jstep + (iz + 0) * kstep;
    int v111 = (ix + 1) * istep + (iy + 1) * jstep + (iz + 1) * kstep;

    float b000 = (1.0 - dx) * (1.0 - dy) * (1.0 - dz);
    float b001 = (1.0 - dx) * (1.0 - dy) * (dz);
    float b010 = (1.0 - dx) * (dy) * (1.0 - dz);
    float b011 = (1.0 - dx) * (dy) * (dz);
    float b100 = (dx) * (1.0 - dy) * (1.0 - dz);
    float b101 = (dx) * (1.0 - dy) * (dz);
    float b110 = (dx) * (dy) * (1.0 - dz);
    float b111 = (dx) * (dy) * (dz);

    if (v->get_type() == Volume::FLOAT)
    {
      float *p = (float *)v->get_samples();
      return p[v000]*b000 + p[v001]*b001 + p[v010]*b010 + p[v011]*b011 +
             p[v100]*b100 + p[v101]*b101 + p[v110]*b110 + p[v111]*b111;
    }
    else
    {
      unsigned char *p = (unsigned char *)v->get_samples();
      return p[v000]*b000 + p[v001]*b001 + p[v010]*b010 + p[v011]*b011 +
             p[v100]*b100 + p[v101]*b101 + p[v110]*b110 + p[v111]*b111;
    }
  }

public:
  bool CollectiveAction(MPI_Comm c, bool s)
  {
    args *a = (args *)contents->get();

    VolumeP v = Volume::Cast(KeyedDataObject::GetByKey(a->vk));
    ParticlesP p = Particles::Cast(KeyedDataObject::GetByKey(a->pk));

    p->CopyPartitioning(v);

    float deltaX, deltaY, deltaZ;
    v->get_deltas(deltaX, deltaY, deltaZ);

    float ox, oy, oz;
    v->get_local_origin(ox, oy, oz);

    int lioff, ljoff, lkoff;
    v->get_ghosted_local_offsets(lioff, ljoff, lkoff);

    int nli, nlj, nlk;
    v->get_local_counts(nli, nlj, nlk);

    int gnli, gnlj, gnlk;
    v->get_ghosted_local_counts(gnli, gnlj, gnlk);

    int local_count = nli*nlj*nlk;

    float max_x = ox + (nli-1) * deltaX,
          max_y = oy + (nlj-1) * deltaY,
          max_z = oz + (nlk-1) * deltaZ;

    int ngx, ngy, ngz;
    v->get_local_counts(ngx, ngy, ngz);
    int global_count = ngx*ngy*ngz;

    istep = 1;
    jstep = gnli;
    kstep = gnli * gnlj;

    Particle tp;
    tp.xyz = get_starting_point(v);
    tp.u.value = sample(v, tp.xyz);

    float tq = Q(v, tp.u.value, a);

    Box *partition_box = v->get_local_box();

    int miss_count = 0;
    for (int iteration = 0; iteration < (a->n_start + a->n_iterations); )
    {
      double rv[3];
      normally_distributed_vector(rv);

      Particle cp(tp.xyz.x + rv[0], tp.xyz.y + rv[1], tp.xyz.z + rv[2], float(0.0));
      if (! partition_box->isIn(cp.xyz)) continue;

      iteration ++;

      cp.u.value = sample(v, cp);
      float cq = Q(v, cp.u.value, a);

      if ((cq > tq) || (RNDM < (cq/tq)))
      {
        tp = cp;
        tq = cq;
        if ((iteration > a->n_start) && ((iteration % a->n_skip) == 0))
          p->push_back(tp);
        miss_count = 0;
      }
      else if (++miss_count > a->n_miss)
      {
        tp.xyz = get_starting_point(v);
        tp.u.value = sample(v, tp.xyz);
        miss_count = 0;
      }
    }

    std::cerr << "created " << p->get_n_samples() << " samples\n";
    return false;
  }
};

WORK_CLASS_TYPE(SampleMsg)

void
syntax(char *a)
{
  cerr << "syntax: " << a << " [options] data" << endl;
  cerr << "optons:" << endl;
  cerr << "  -D            run debugger" << endl;
  cerr << "  -M m          max number of consecutive misses allowed (" << max_consecutive_misses << ")" << "\n";
  cerr << "  -i n          iteration limit (20000)\n";
  cerr << "  -h n          hot start count (1000)\n";
  cerr << "  -S step       scaling of candidate step (1)\n";
  cerr << "  -K skip       record every skip'th successful sample (10)\n";
  cerr << "  -R p0 p1      linear transfer function, from p0 to p1 extended (default, linear data range)\n";
  cerr << "  -G p0 p1      gaussian transfer function p0 is median, p2 is stdev (default, linear data range)\n";
  cerr << "  -s x y        window size (" << WIDTH << "x" << HEIGHT << ")" << endl;
  cerr << "  -t target     value of highest interest (" << target << ")" << endl;
  cerr << "  -p power      exponent of interest (" << power << ")" << endl;
  cerr << "  -r radius     radius of samples (" << radius << ")" << endl;
  cerr << "  -k sigma      standard deviation of normally distributed random step vector (" << sigma << ")" << endl;
  exit(1);
}

#if 0
namespace gxy
{
extern void DDSpheres_Hello();
};
#endif

int
main(int argc, char * argv[])
{
  string data = "";
  char *dbgarg;
  bool dbg = false;

#if 0
  dbg = true;
  std::cerr << getpid() << "\n";
  while (dbg)
    sleep(1);
#endif

  ospInit(&argc, (const char **)argv);

#if 0
  DDSpheres_Hello();

  {
    void *foo = dlopen(NULL, RTLD_NOW | RTLD_GLOBAL);
    void *s = dlsym(foo, "ospray_init_module_ispc");
    if (s) std::cerr << "ospray_init_module_ispc OK\n"; else std::cerr << "ospray_init_module_ispc NOT FOUND\n";

    s = dlsym(foo, "ospray_create_geometry__dspheres");
    if (s) std::cerr << "ospray_create_geometry__dspheres OK\n"; else std::cerr << "ospray_create_geometry__dspheres NOT FOUND\n";
  }
#endif

  Application theApplication(&argc, &argv);
  theApplication.Start();

  for (int i = 1; i < argc; i++)
  {
    if (argv[i][0] == '-')
    {
      switch (argv[i][1])
      {
        case 'D': dbg = true, dbgarg = argv[i] + 2; break;
        case 'M': max_consecutive_misses = atoi(argv[++i]); break;
        case 'i': iteration_limit = atoi(argv[++i]); break;
        case 'h': hot_start = atoi(argv[++i]); break;
        case 'S': scaling = atof(argv[++i]); break;
        case 'K': skip = atoi(argv[++i]); break;
        case 'R': tf0 = atof(argv[++i]), tf1 = atof(argv[++i]), tf_type = TF_LINEAR; break;
        case 'G': tf0 = atof(argv[++i]), tf1 = atof(argv[++i]), tf_type = TF_GAUSSIAN; break;
        case 'r': radius = atof(argv[++i]); break;
        case 't': target = atof(argv[++i]); break;
        case 'p': power  = atof(argv[++i]); break;
        case 'k': sigma = atof(argv[++i]); break;
        case 's': width  = atoi(argv[++i]); height = atoi(argv[++i]); break;
        default:
          syntax(argv[0]);
      }
    }
    else if (data == "") data = argv[i];
    else syntax(argv[0]);
  }

  generator = new variate_generator<mt19937, normal_distribution<> >(mt19937(time(0)), normal_distribution<>(0.0, sigma));

  Renderer::Initialize();
  theApplication.Run();

  RendererP theRenderer = Renderer::NewP();

  mpiRank = theApplication.GetRank();
  mpiSize = theApplication.GetSize();

  srand(mpiRank);

  Debug *d = dbg ? new Debug(argv[0], false, dbgarg) : NULL;

  SampleMsg::Register();

  if (mpiRank == 0)
  {
    theRenderer->Commit();

    VolumeP volume = Volume::NewP();
    volume->Import(data);

    if (tf_type == TF_NONE)
      volume->get_global_minmax(tf0, tf1), tf_type = TF_LINEAR;

    ParticlesP samples = Particles::NewP();
    samples->SetRadius(radius);

    SampleMsg *smsg = new SampleMsg(volume, samples, tf_type, tf0, tf1, scaling, iteration_limit, hot_start, skip, max_consecutive_misses);
    smsg->Broadcast(true, true);

    samples->Commit();

    DatasetsP theDatasets = Datasets::NewP();
    theDatasets->Insert("samples", samples);
    theDatasets->Insert("volume", volume);
    theDatasets->Commit();

    vector<CameraP> theCameras;

#if 1
    for (int i = 0; i < 20; i++)
    {
      CameraP cam = Camera::NewP();

      cam->set_viewup(0.0, 1.0, 0.0);
      cam->set_angle_of_view(30.0);

      float angle = 2*3.1415926*(i / 20.0);

      float vpx = 8.0 * cos(angle);
      float vpy = 8.0 * sin(angle);

      cam->set_viewpoint(vpx, vpy, 0.0);
      cam->set_viewdirection(-vpx, -vpy, 0.0);

      cam->Commit();
      theCameras.push_back(cam);
    }
#else
    CameraP cam = Camera::NewP();
    cam->set_viewup(0.0, 1.0, 0.0);
    cam->set_angle_of_view(45.0);
    cam->set_viewpoint(4.0, 0.0, 0.0);
    cam->set_viewdirection(-2.0, 0.0, 0.0);
    cam->Commit();
    theCameras.push_back(cam);
#endif

    ParticlesVisP pvis = ParticlesVis::NewP();
    pvis->SetName("samples");
    pvis->Commit(theDatasets);

    VolumeVisP vvis = VolumeVis::NewP();
    vvis->SetName("volume");

    vec4f cmap[] = {
        {0.00,1.0,0.5,0.5},
        {0.25,0.5,1.0,0.5},
        {0.50,0.5,0.5,1.0},
        {0.75,1.0,1.0,0.5},
        {1.00,1.0,0.5,1.0}
    };

    vec2f omap[] = {
                { 0.00, 0.00},
                { 1.00, 0.10},
                { 8.00, 0.10}
               };

    vvis->SetColorMap(5, cmap);
    vvis->SetOpacityMap(3, omap);
    vvis->SetVolumeRendering(true);
    vvis->Commit(theDatasets);

    VisualizationP v = Visualization::NewP();
    float light[] = {1.0, 2.0, 3.0}; int t = 1;
    v->get_the_lights()->SetLights(1, light, &t);
    v->get_the_lights()->SetK(0.4, 0.6);
    v->get_the_lights()->SetShadowFlag(false);
    v->get_the_lights()->SetAO(0, 0.0);
    v->AddVis(pvis);
    v->AddVis(vvis);
    v->Commit(theDatasets);

    RenderingSetP theRenderingSet = RenderingSet::NewP();

    int indx = 0;
    for (auto c : theCameras)
    {
      RenderingP theRendering = Rendering::NewP();
      theRendering->SetTheOwner((indx++) % mpiSize);
      theRendering->SetTheSize(width, height);
      theRendering->SetTheCamera(c);
      theRendering->SetTheDatasets(theDatasets);
      theRendering->SetTheVisualization(v);
      theRendering->Commit();
      theRenderingSet->AddRendering(theRendering);
    }

    theRenderingSet->Commit();

    theRenderer->Render(theRenderingSet);

#ifdef GXY_WRITE_IMAGES
    theRenderingSet->WaitForDone();
#endif
    
    theRenderingSet->SaveImages(string("samples"));

    theApplication.QuitApplication();
  }

  theApplication.Wait();
}
